Diggy - New leveling system

SimonFlapse opened the pull request:
Transition of old Diggy level system to new based on ForceControl.lua
Allows removing stone requirements.

- [x] Setup basic control

- [x] Transfer formula

- [x] Handle awarding experience
    - Mining stone (Needs tweaking)
    - Killing biters/worms (Needs tweaking)
    - Research? (Needs tweaking. All research, except for infinity, gives almost 5 million XP)
    - Rocket launches? (Needs tweaking, and testing)

- [x] Implement buffs
    - Mining boost
    - Inventory
    - Health

- [x] Implement market items unlocking

- [x] Implement GUI interface

- [x] Change market prices (Initial - Needs tweaking)

- [x] Documentation

- [x]  DiggyHole.lua testing (Removed inventory full check)

- [x] Write CHANGELOG.md

SimonFlapse committed:
Basic setup for Force Control

Does nothing else than give experience for each mined stone.

linaori commented on features/force_control.lua:
```diff
@@ -160,7 +160,7 @@ function ForceControl.register_force(lua_force_or_name)
         current_experience = 0,
         current_level = 0,
-        experience_level_up_cap = next_level_cap_calculator.execute(0),
+        experience_level_up_cap = next_level_cap_calculator.execute(1),
```
Current level is 0, so if you want to start level 1, the current_level should also be 1. Either way is fine for me, but they should be kept in sync I think

SimonFlapse replied:
Yeah, I'm just having a problem with leveling up. Since level 0's cap is at 25 and not 50 with my formula. It seems that it calculates the upper bound of a level. But I know how to go around this now.

linaori commented on features/force_control.lua:
```diff
@@ -257,4 +257,18 @@ function ForceControl.get_force_data(lua_force_or_name)
+        return
+    end
+
+    return 'Current experience: ' .. force_config.current_experience .. ' Current level: ' .. force_config.current_level .. ' Next level at: ' .. force_config.experience_level_up_cap .. ' Percentage to level up: ' .. ((force_config.current_experience / force_config.experience_level_up_cap) * 100)
```
`force_config.experience_percentage` can also be used (it's calculated in the function above)

SimonFlapse replied:
It's not inside that scope. Just returns nil

linaori replied:
So what if you do:
```lua
function ForceControl.get_formatted_force_data(lua_force_or_name)
    local force_config = ForceControl.get_force_data(lua_force_or_name)
    if not force_config then
        return
    end

    return 'Current experience: ' .. force_config.current_experience .. ' Current level: ' .. force_config.current_level .. ' Next level at: ' .. force_config.experience_level_up_cap .. ' Percentage to level up: ' .. force_config.experience_percentage .. '%'
end
```

linaori commented on map_gen/Diggy/Feature/Experience.lua:
```diff
@@ -0,0 +1,57 @@
+local log10 = math.log10
+local force = 'player'
+
+function Experience.calculate_level(level) -- all configurable variables must be integers.
```
I don't think this function has to be publicly available

linaori commented on map_gen/Diggy/Feature/Experience.lua:
```diff
@@ -0,0 +1,57 @@
+    if ((entity.name ~= 'sand-rock-big') and (entity.name ~= 'rock-huge')) then
+        return
+    end
+    local exp = math.random(19,24)
```
Can we make this a stable average number? As the player doesn't loot stone, we can also show this in a message via one of the Game print floating text functions. You can use it to show `+x experience`. Would show it at the rock, not the player. The same can be done for killing biters.

SimonFlapse replied:
Gotcha. So no random experience for anything

SimonFlapse committed:
Buffs transitioned to new level system

Implemented level up buffs (Also made a double value avaliable at certain levels)

Added health bonus

Added experience for research, rocket launched, mined rocks

Removing experience on player death. 0.5% of total experience

Added total_experience to ForceControl.lua

Refactored util/game.lua

linaori commented on features/force_control.lua:
```diff
@@ -82,8 +82,9 @@ function ForceControlBuilder.register(level_matches, callback, lua_force_name)
-        if level_matches(level, event.force) then
-            callback(level, event.force)
+        local force = get_valid_force(event.force)
+        if level_matches(level, force) then
```
Won't this prevent the `LuaForce` from reaching the callback? Meaning you can't directly call it anymore

SimonFlapse replied:
It's not meant to be there, I used it to test why I couldn't add buffs to the force.

linaori commented on map_gen/Diggy/Feature/Experience.lua:
```diff
@@ -0,0 +1,233 @@
+function Experience.on_init()
+    local force = game.forces.player
+    ForceControl.register_force(force)
+    log(force.name)
```
Don't forget 😄

linaori reviewed (commented):
Looking good so far 😄

SimonFlapse committed:
Transfered market item unlocks

Made Market Items unlock through the force control system.

Made things configurable from the Config.lua

Removed stone from mined stone

SimonFlapse committed:
Initial conversion of market items to coins

Needs testing

SimonFlapse committed:
Started biter killing giving XP

SimonFlapse committed:
Merge branch 'develop' of https://github.com/SimonFlapse/RedMew into Diggy-ForceControl-Level

SimonFlapse committed:
XP for killing enemies. GUI update

linaori commented on map_gen/Diggy/Feature/Experience.lua:
```diff
@@ -101,12 +100,9 @@ local function on_player_mined_entity(event)
-        exp = 10
+    local exp
+    if entity.name == 'sand-rock-big' or entity.name == 'rock-huge' then
+        exp = Config.XP['' .. entity.name]
```
This looks a bit like an odd structure. What about `Config.XP[entity.name]` instead? Another option would be this:
```lua
local sand_rock_xp = Config.XP['sand-rock-big']
local huge_rock_xp = Config.XP['rock-huge']
local function on_player_mined_entity(event)
    if entity.name == 'sand-rock-big' then
        exp = sand_rock_xp
    end
    if entity.name == 'rock-huge' then
        exp = huge_rock_xp
    end
end
```
This has the benefit of no deep index lookup when it dies (500k+ in a big map). If you still have issues with the floating text, we could also drop coins instead of award them.

SimonFlapse replied:
Yeah you're right, I'll define all these configurable XP on register to prevent all the table lookups for the exact same value.

linaori commented on map_gen/Diggy/Feature/Experience.lua:
```diff
@@ -176,23 +157,23 @@ local function on_rocket_launched(event)
 
     ForceControl.get_force_data(force)
-
-    local exp = ForceControl.get_force_data(force).total_experience*0.005
```
What if we add a function in `ForceControl` for this?
```lua
local exp_removed = ForceControl.remove_experience_percentage(death_penalty)
```

SimonFlapse replied:
Alright :) Generic function anyways

linaori reviewed (commented):
Can you add 'stone' to the item removed when mining in DiggyHole as well? (look for 'coal' and you'll find it)

SimonFlapse committed:
Refactoring of remove experience percentage to generic force_control

linaori commented on features/force_control.lua:
```diff
@@ -200,8 +201,33 @@ function ForceControl.remove_experience(lua_force_or_name, experience)
+    if not force_config then
+        return
+    end
+    local ceil = math.ceil
```
Can push this to the top of the file, it will only assign the reference once instead of every time it enters this scope

linaori commented on features/force_control.lua:
```diff
@@ -200,8 +201,33 @@ function ForceControl.remove_experience(lua_force_or_name, experience)
+    force_config.total_experience = (force_config.current_experience == 0) and force_config.total_experience - backup_current_experience or max(0, force_config.total_experience - experience)
+end
+
+---Removes experience from a force, based on a percentage of the total experience
```
total experience obtained, or total experience needed to level up?

SimonFlapse replied:
Total experience obtained

linaori commented on features/force_control.lua:
```diff
@@ -200,8 +201,33 @@ function ForceControl.remove_experience(lua_force_or_name, experience)
+
+    local penalty = force_config.total_experience * percentage
+    penalty = (penalty >= min_experience) and ceil(penalty) or ceil(min_experience)
+    ForceControl.remove_experience(lua_force_or_name, penalty)
```
What if we move the removed XP to the remove_experience function? Then we could use the same functionality for that scope and just return the actual XP removed directly

SimonFlapse replied:
Not sure I fully understand, The reason i'm returning the value is that I needed it to display the floating text. And since you'd normally supply it with experience to add or remove you'd already know how much that is, using the other functions. The ceil part is because it's a percentage and I'd like to avoid subtracting weird fractions.

linaori replied:
I'm referring to doing this: `return ForceControl.remove_experience(lua_force_or_name, penalty)`. This function already calculates the difference between how much you wanted to subtract and how much you _can_ subtract until your reach 0. If you let that function return that number, you only have to return that in here and not worry about the difference between going from (say) 25 xp - 50 to 0 or -25.

SimonFlapse replied:
Okay I understand now. Yeah I'll like that. That way I'm not relaying wrong information to the players via the floating text if we can't remove all the experience

linaori commented on map_gen/Diggy/Config.lua:
```diff
@@ -298,7 +298,7 @@ local Config = {
             market_spawn_position = {x = 0, y = 3},
             stone_to_surface_amount = 50,
-            currency_item = 'stone',
+            currency_item = 'coin',
```
I think we can hard-code this one by now 😄

SimonFlapse replied:
Doesn't hurt to give people the opportunity to change it?

linaori replied:
It's fine either way, just makes it close to useless as it wasn't designed for it. Where's it actually still used? 🤔

SimonFlapse replied:
It's used when you don't specify what the price of an item is in, it'll automatic assign it to currency_item

linaori commented on map_gen/Diggy/Config.lua:
```diff
@@ -310,52 +310,42 @@ local Config = {
+                    {level = 58, price = 50, name = 'rocket'},
+                    {level = 66, price = 100, name = 'explosive-rocket'},
+                    {level = 73, price = 2000, name = 'satellite'},
+                    {level = 100, price = 1, name = 'iron-stick'},
```
😂

SimonFlapse replied:
Oh yeah, I needed something for level 100 🗡

linaori commented on map_gen/Diggy/Feature/DiggyHole.lua:
```diff
@@ -254,6 +254,7 @@ function DiggyHole.register(config)
+            event.buffer.remove({name = 'stone', count = 100})
 
             -- this logic can be replaced once we've fully replaced the stone to surface functionality
             if enable_digging_warning then
```
Can this warning be removed now? I think we don't use it anywhere anymore once this feature is finished

SimonFlapse replied:
It very much can be removed :)

SimonFlapse replied:
I'd need you to take look at DiggyHole.lua it see if I removed something wrong

linaori commented on map_gen/Diggy/Feature/Experience.lua:
```diff
@@ -0,0 +1,247 @@
+    local buff = Config.buffs['mining_speed']
+    if level_up > 0 and buff ~= nil then
+        local value = (buff.double_level ~= nil and level_up%buff.double_level == 0) and buff.value*2 or buff.value
+        mining_efficiency.level_modifier = mining_efficiency.level_modifier + (value / 100)
```
Minor thing so it won't matter much, but it's interesting to do nevertheless. In lua multiplication is faster than division, so you can also do `(value * 0.01)`: http://lua-users.org/wiki/OptimisationCodingTips

Performance is probably neglect-able though

SimonFlapse committed:
LuaDoc

linaori reviewed (commented):
Looking good so far, make sure to not forget to test it in heavy mode

SimonFlapse committed:
Merge branch 'develop' of https://github.com/SimonFlapse/RedMew into Diggy-ForceControl-Level

SimonFlapse committed:
Minor improvements, disabled nukes, removed inventory_full_warning from DiggyHole

linaori commented on map_gen/Diggy/Feature/Experience.lua:
```diff
@@ -164,6 +168,8 @@ local function on_research_finished(event)
     Experience.update_inventory_slots(force, 0)
     Experience.update_mining_speed(force, 0)
+
+    game.forces.player.technologies['landfill'].enabled = false
```
Could be moved up to where the atomic bomb is disabled (and use `force.technologies`

SimonFlapse replied:
But we don't need to set that more than once? The other it just to disable it when the research is researched. I could implement a check for if the nuclear research was the one being researched and only disable it then. But the current solution works :D

linaori reviewed (approved):
I think it's almost if not completely ready to be merged?

SimonFlapse committed:
Made ForceControl.remove_experience return the actual removed experience

SimonFlapse committed:
Got my head and mind back, also fixed progress bar saying long numbers

SimonFlapse commented:
Fails heavy-mode

SimonFlapse commented:
This is causing me to fail heavy-mode. It's from the old diggy market (Seems to spawn it after 50 ticks). Removing it makes me pass heavy-mode
```lua
local on_market_timeout_finished = Token.register(function(params)
    Template.market(params.surface, params.position, params.player_force, {})
end)
```
```lua
function MarketExchange.on_init()
    Task.set_timeout_in_ticks(50, on_market_timeout_finished, {
        surface = game.surfaces.nauvis,
        position = config.market_spawn_position,
        player_force = game.forces.player,
    })
end
```
Okay so Template.market raises an event that triggers this function:
```lua
local function on_placed_entity(event)
    local market = event.entity
    if 'market' ~= market.name then
        return
    end
    diggy_market = market
end
```
It seems that me assigning the market to the local variable diggy_market causes the error

SimonFlapse commented:
Researching (all) technologies causes a failure in heavy mode too

iltar commented:
@Valansch would this issue be fixed by using a token instead of a global inside the force control?

SimonFlapse committed:
Changed get_market()

SimonFlapse committed:
Workaround for desync

SimonFlapse commented:
Made a workaround, it passes heavy mode now

SimonFlapse committed:
Forgot to remove a debug

linaori commented on map_gen/Diggy/Feature/Experience.lua:
```diff
@@ -0,0 +1,296 @@
+local function on_entity_died (event)
+    local entity = event.entity
+    local force = entity.force
+    local exp = 0
```
local assignment can go below the force.name check

linaori commented on map_gen/Diggy/Feature/Experience.lua:
```diff
@@ -0,0 +1,296 @@
+---@param level number a number specifying the level
+---@return number required total experience to reach supplied level
+function Experience.calculate_level_xp(level)
+        if level_table[level] == nil then
```
indent

linaori reviewed (approved):
Looks good!

Valansch committed:
[Diggy] Experience minor refactoring

Valansch commented:
@stickler-ci shut up :D

Valansch committed:
fixed nil entity

plague006 commented:
Rebase the PR and let the bot do its thing :p

SimonFlapse committed:
Merge branch 'develop' of https://github.com/Refactorio/RedMew into Diggy-ForceControl-Level

SimonFlapse commented on map_gen/Diggy/Feature/Experience.lua:
```diff
@@ -0,0 +1,295 @@
+    ForceControl_builder.register_on_every_level(function (level_reached, force)
+        force.print(string_format('Leved up to %d!', level_reached))
+        force.play_sound{path='utility/new_objective', volume_modifier = 1 }
+        local Experience = require 'map_gen.Diggy.Feature.Experience'
```
This is required for the function to work

plague006 commented:
Don't take stickler's warnings as gospel. If something is harmless and you just don't have the time or energy to change it: it's fine.
Things like defining non-standard globals in particular

Valansch commented on map_gen/Diggy/Feature/Experience.lua:
```diff
@@ -0,0 +1,295 @@
+    if not cause or cause.type ~= 'player' or not cause.valid then
+        return
+    end
+    local exp = 0
```
My bad, this can be removed

SimonFlapse commented on features/force_control.lua:
```diff
@@ -200,8 +199,32 @@ function ForceControl.remove_experience(lua_force_or_name, experience)
+---@return number the experience being removed
+---@see ForceControl.remove_experience
+function ForceControl.remove_experience_percentage(lua_force_or_name, percentage, min_experience)
+    local min_experience = min_experience ~= nil and min_experience or 0
```
It's supposed to be shadowed

SimonFlapse committed:
Brand new formula, CI review changes

SimonFlapse committed:
Fixed setting an value that got overridden before use

SimonFlapse committed:
Merge branch 'develop' of https://github.com/Refactorio/RedMew into Diggy-ForceControl-Level

SimonFlapse committed:
Update CHANGELOG.md
